sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordWriteSchemaTransformProvider.java

```diff
@@ -0,0 +1,268 @@
+  private static final String INPUT = "input";
+  private static final String OUTPUT = "output";
+  private static final String ERROR = "errors";
+  public static final TupleTag<byte[]> OUTPUT_TAG = new TupleTag<byte[]>() {};
+  public static final TupleTag<Row> ERROR_TAG = new TupleTag<Row>() {};
+  private static final Logger LOG =
+      LoggerFactory.getLogger(TFRecordWriteSchemaTransformProvider.class);
+
+  /** Returns the expected class of the configuration. */
+  @Override
+  protected Class<TFRecordWriteSchemaTransformConfiguration> configurationClass() {
+    return TFRecordWriteSchemaTransformConfiguration.class;
+  }
+
+  /** Returns the expected {@link SchemaTransform} of the configuration. */
+  @Override
+  protected SchemaTransform from(TFRecordWriteSchemaTransformConfiguration configuration) {
+    return new TFRecordWriteSchemaTransform(configuration);
+  }
+
+  /** Implementation of the {@link TypedSchemaTransformProvider} identifier method. */
+  @Override
+  public String identifier() {
+    return IDENTIFIER;
+  }
+
+  /** Implementation of the {@link TypedSchemaTransformProvider} inputCollectionNames method. */
+  @Override
+  public List<String> inputCollectionNames() {
+    return Arrays.asList(INPUT, ERROR);
+  }
+
+  /** Implementation of the {@link TypedSchemaTransformProvider} outputCollectionNames method. */
+  @Override
+  public List<String> outputCollectionNames() {
+    return Arrays.asList(OUTPUT, ERROR);
+  }
+
+  /**
+   * An implementation of {@link SchemaTransform} for TFRecord Write jobs configured using {@link
+   * TFRecordWriteSchemaTransformConfiguration}.
+   */
+  static class TFRecordWriteSchemaTransform extends SchemaTransform {
+    private final TFRecordWriteSchemaTransformConfiguration configuration;
+
+    TFRecordWriteSchemaTransform(TFRecordWriteSchemaTransformConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    public Row getConfigurationRow() {
+      try {
+        // To stay consistent with our SchemaTransform configuration naming conventions,
+        // we sort lexicographically
+        return SchemaRegistry.createDefault()
+            .getToRowFunction(TFRecordWriteSchemaTransformConfiguration.class)
+            .apply(configuration)
+            .sorted()
+            .toSnakeCase();
+      } catch (NoSuchSchemaException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public PCollectionRowTuple expand(PCollectionRowTuple input) {
+      // Validate configuration parameters
+      configuration.validate();
+
+      // Create basic transform
+      TFRecordIO.Write writeTransform =
+          TFRecordIO.write().withCompression(Compression.valueOf(configuration.getCompression()));
+
+      // Add more parameters if not null
+      String outputPrefix = configuration.getOutputPrefix();
+      if (outputPrefix != null) {
```

Comment:
nit: no need for these null checks if the SchemaTransformConfiguration object does not mark them as `@Nullable`

Reply:
Should we make it required? I see TFRecordIO makes a null check during expansion: https://github.com/apache/beam/blob/b1f9f2e263203f76e9de683de98500296fb74e00/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java#L392-L395

Reply:
to be fair, TFRecordIO.Write also makes it `@Nullable`: https://github.com/apache/beam/blob/b1f9f2e263203f76e9de683de98500296fb74e00/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java#L259-L260

But if we're gonna fail at expansion time we might as well just make it required? unless i'm missing something